ephyviewer/datasource/epochs.py

```diff
@@ -54,152 +55,219 @@ class WritableEpochSource(InMemoryEpochSource):
-        if ind>0:
-            t2_prev = ep_times[prev]+ep_durations[prev]
-            if (t2_prev)>t2:
-                ep_times = insert_item(ep_times, ind+1, t2)
-                ep_durations = insert_item(ep_durations, ind+1, t2_prev - t2)
-                ep_labels = insert_item(ep_labels, ind+1, ep_labels[prev])
-        #short prev durations
-        while prev>=0:
-            if (ep_times[prev]+ep_durations[prev])>ep_times[ind]:
-                ep_durations[prev] = ep_times[ind] - ep_times[prev]
-            else:
-                break
-            prev = prev-1
-        
-        #nexts
-        next = ind+1
-        while next<ep_times.size:
-            delta = (ep_times[ind]+ep_durations[ind]) - ep_times[next]
-            if delta>0:
-                ep_times[next] += delta
-                ep_durations[next] -= delta
-            else:
-                break
-            next = next + 1
-        
-        self._clean_and_set(ep_times, ep_durations, ep_labels)
+        
+        ep_times, ep_durations, ep_labels, ep_ids = self.ep_times, self.ep_durations, self.ep_labels, self.ep_ids
+        ep_times = np.append(ep_times, t1)
+        ep_durations = np.append(ep_durations, duration)
+        ep_labels = np.append(ep_labels, label)
+        ep_ids = np.append(ep_ids, self._next_id)
+        self._next_id += 1
+        
+        self._clean_and_set(ep_times, ep_durations, ep_labels, ep_ids)
+    
+    def delete_epoch(self, ind):
+        
+        ep_times, ep_durations, ep_labels, ep_ids = self.ep_times, self.ep_durations, self.ep_labels, self.ep_ids
+        ep_times = np.delete(ep_times, ind)
+        ep_durations = np.delete(ep_durations, ind)
+        ep_labels = np.delete(ep_labels, ind)
+        ep_ids = np.delete(ep_ids, ind)
+        
+        self._clean_and_set(ep_times, ep_durations, ep_labels, ep_ids)
     
     def delete_in_between(self, t1, t2):
-        #~ print('delete_in_between', t1, t2)
-        ep_times, ep_durations, ep_labels = self.all[0]['time'], self.all[0]['duration'], self.all[0]['label']
-        
-        ep_stops = ep_times+ep_durations
-        
-        i1 = np.searchsorted(ep_times, t1, side='left')
-        i2 = np.searchsorted(ep_times+ep_durations, t2, side='right')
-        #~ print 'i1, i2', i1, i2
-        #~ print()
-        for i in range(i1-1, i2+1):
-            #~ print i
-            if i1<0: continue
-            if i>=ep_durations.size: continue
+        
+        ep_times, ep_durations, ep_stops, ep_labels, ep_ids = self.ep_times, self.ep_durations, self.ep_stops, self.ep_labels, self.ep_ids
+        
+        for i in range(len(ep_times)):
             
-            if ep_times[i]>=t1 and ep_stops[i]<t2:
-                #~ print 'a'
-                ep_durations[i] = -1.
-            elif ep_times[i]<t1 and (t1<ep_stops[i]<t2):
-                #~ print 'b'
+            # if epoch starts and ends inside range, delete it
+            if ep_times[i]>=t1 and ep_stops[i]<=t2:
```

Comment:
**ep_stops[i]<=t2** became **ep_stops[i]<t2**.

I don't remember why it was important for me.
The general python approach is than the left limit is included (>=) and the right limit is excluded (<).

pandas apporach with .loc[t1:t2] with float index this your one. (right limit included).

Did you choose that intionally ?

Reply:
Yes, it was intentional and necessary to fix a bug that existed before I began this PR. Epochs would fail to delete with the region deletion tool if they started within the selected region and ended _exactly_ at its right boundary. None of the `if` clauses below this would catch that case. This case was probably not noticed before rounding of times to the nearest microsecond was added simply due to small numerical inaccuracy in floating point numbers.